internal/impl/snowflake/output_snowflake_streaming.go

```diff
@@ -533,49 +717,130 @@ func (o *snowflakeStreamerOutput) WriteBatch(ctx context.Context, batch service.
 		}
 		batch = mapped
 	}
-	if !o.hasOpenedChannels() {
-		if len(batch) == 0 {
-			return nil
-		}
-		if err := o.CreateOutputTable(ctx, batch); err != nil {
-			return err
-		}
-		// We're not connected so tell the framework to call connect to re-connect to open our channel
-		return service.ErrNotConnected
-	}
 	var err error
 	// We only migrate one column at a time, so tolerate up to 10 schema
 	// migrations for a single batch before giving up. This protects against
 	// any bugs over infinitely looping.
 	for i := 0; i < 10; i++ {
-		err = o.WriteBatchInternal(ctx, batch)
+		o.mu.RLock()
+		err = o.impl.WriteBatch(ctx, batch)
+		o.mu.RUnlock()
 		if err == nil {
 			return nil
 		}
-		migrationErr := schemaMigrationNeededError{}
-		if !errors.As(err, &migrationErr) {
-			break
+		if o.schemaEvolver == nil {
+			return err
+		}
+		if streaming.IsTableNotExistsError(err) {
+			if err := o.createTable(ctx, batch); err != nil {
+				return err
+			}
+			continue // If creating the table succeeded, retry
 		}
-		if err := migrationErr.migrator(ctx); err != nil {
+		needsMigrationErr := schemaMigrationNeededError{}
+		if !errors.As(err, &needsMigrationErr) {
 			return err
 		}
+		o.mu.Lock()
+		migrateErr := o.runMigration(ctx, needsMigrationErr)
+		o.mu.Unlock()
+		if migrateErr != nil {
+			return migrateErr
+		}
 	}
 	return err
 }
 
-func (o *snowflakeStreamerOutput) WriteBatchInternal(ctx context.Context, batch service.MessageBatch) error {
-	o.schemaMigrationMu.RLock()
-	defer o.schemaMigrationMu.RUnlock()
+func (o *snowpipeStreamingOutput) createTable(ctx context.Context, batch service.MessageBatch) error {
+	o.mu.Lock()
+	defer o.mu.Unlock()
+	if err := o.schemaEvolver.CreateOutputTable(ctx, batch); err != nil {
+		return err
+	}
+	if err := o.impl.Connect(ctx); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (o *snowpipeStreamingOutput) runMigration(ctx context.Context, needsMigrationErr schemaMigrationNeededError) error {
+	if err := needsMigrationErr.runMigration(ctx, o.schemaEvolver); err != nil {
+		return err
+	}
+	// After a migration we need to reopen all our channels
+	// so close and reopen our impl
+	if err := o.impl.Close(ctx); err != nil {
```

Comment:
Is this safe when `max_in_flight` is greater than 1?

Reply:
Yes we're holding a lock